dLux/apertures.py

```diff
@@ -0,0 +1,2641 @@
+import equinox as eqx
+import matplotlib as mpl 
+import matplotlib.pyplot as plt
+import jax.numpy as np
+import jax 
+import dLux
+import abc
+import functools
+from typing import TypeVar 
+
+
+Array = np.ndarray
+Wavefront = dLux.wavefronts.Wavefront
+
+
+__all__ = ["factorial", "CircularAperture", "SquareAperture", 
+    "HexagonalAperture", "RegularPolygonalAperture", 
+    "IrregularPolygonalAperture", "StaticAperture",
+    "AberratedAperture", "StaticAberratedAperture"]
+
+
+two_pi: float = 2. * np.pi
+        
+ 
+def factorial(n : int) -> int:
+    """
+    Calculate n! in a jax friendly way. Note that n == 0 is not a 
+    safe case.  
+ 
+    Parameters
+    ----------
+    n : int
+        The integer to calculate the factorial of.
+ 
+    Returns
+    n! : int
+        The factorial of the integer
+    """
+    return jax.lax.exp(jax.lax.lgamma(n + 1.))
+
+
+
+class ApertureLayer(dLux.optics.OpticalLayer, abc.ABC):
+    """
+    The ApertureLayer groups together all of the functionality 
+    that is associated with the apertures. Very little of this
+    functionality is actually implemented by itself because 
+    implementation varies amongst the subclasses. It is a 
+    layer within the class Heirachy that exists almost purely 
+    as a classification.
+
+    Parameters:
+    -----------
+    name: String
+        The address of this ApertureLayer within the optical 
+        system. 
+    """
+
+    
+    def __init__(self: dLux.optics.OpticalLayer) -> dLux.optics.OpticalLayer:
+        """
+        Automatically assigns the name of the layer to be the 
+        class name. 
+        """
+        self.name = self.__class__.__name__
```

Comment:
This is a nice little trick, I have a q though: Will this instantiate the name of the lowest or highest level class? Ie will this become `HexagonalAperture` or `ApertureLayer` one this is inherited?